mivideo/mivideo_extract.py

```diff
@@ -0,0 +1,134 @@
+# -*- coding: utf-8 -*-
+'''
+Module for setting up and running the MiVideo data extract.
+'''
+
+import logging
+import os
+import time
+from typing import Union, Sequence, Dict
+
+import pandas as pd
+from google.cloud import bigquery
+from google.oauth2 import service_account
+from sqlalchemy.engine import ResultProxy
+
+import mivideo.queries as queries
+from db.db_creator import DBCreator
+from environ import ENV, CONFIG_DIR
+from vocab import ValidDataSourceName
+
+logger = logging.getLogger(__name__)
+
+SHAPE_ROWS: int = 0  # Index of row count in DataFrame.shape() array
+
+
+class MiVideoExtract(object):
+    '''
+    Initialize the MiVideo data extract process by instantiating the ``MiVideoExtract`` class,
+    then invoke its ``run()`` method.
+
+    For example, ``MiVideoExtract().run()``.
+    '''
+
+    def __init__(self):
+        self.udpKeyFilePath: str = os.path.join(CONFIG_DIR, ENV.get('mivideo', {}).get(
+            'service_account_json_filename'))
+        logger.debug(f'udpKeyFilePath: "{self.udpKeyFilePath}"')
+
+        self.credentials: service_account.Credentials = \
+            service_account.Credentials.from_service_account_file(
+                self.udpKeyFilePath,
+                scopes=['https://www.googleapis.com/auth/cloud-platform'],
+            )
+
+        self.udpDb: bigquery.Client = bigquery.Client(
+            credentials=self.credentials,
+            project=self.credentials.project_id,
+        )
+
+        logger.info(f'Connected to BigQuery project: "{self.udpDb.project}"')
+
+        DB_PARAMS = ENV['INVENTORY_DB']
+        APPEND_TABLE_NAMES = ENV.get('APPEND_TABLE_NAMES', [
+            'job_run', 'data_source_status', 'mivideo_media_started_hourly',
+            'mivideo_media_creation'])
+
+        self.appDb: DBCreator = DBCreator(DB_PARAMS, APPEND_TABLE_NAMES)
+        self.appDb.set_up()
+
+    def _readTableLastTime(self, tableName: str, tableColumnName: str) -> Union[str, None]:
+        lastTime: Union[str, None]
+
+        try:
+            sql: str = f'select max(t.{tableColumnName}) from {tableName} t'
+            result: ResultProxy = self.appDb.engine.execute(sql)
+            lastTime = str(result.fetchone()[0])
+        except(Exception):
+            lastTime = None
```

Comment:
When I ran this for the first time, there was no exception caught and instead the value of `lastTime` was the string `"None"`, which led to an invalid timestamp being passed to the query. I fixed it by changing to the following:

```
        sql: str = f'select max(t.{tableColumnName}) from {tableName} t'
        result: ResultProxy = self.appDb.engine.execute(sql)
        lastTime = str(result.fetchone()[0])

        if lastTime == 'None':
            lastTime = None
```

Reply:
Will check…

Reply:
Yeah, I was thinking about my change later that added `str()` there.  It's probably more useful for the function to return a time-like object (or `None`) as it was rather than `str`.  I needed to use the return value as a string somewhere else, so I should have cast it as such at that point, not make this function return `str`.

Reply:
Just because I was doing a substitution into a SQL string doesn't mean that the value needed to be a string, too.  I changed everything to use `datetime`.